ouroboros-consensus/src/Ouroboros/Consensus/Util/Counting.hs

```diff
@@ -66,6 +72,16 @@ import           Ouroboros.Consensus.Util.SOP
 
 newtype Exactly xs a = Exactly { getExactly :: NP (K a) xs }
 
+instance (SerialiseNodeToClient blk a, SListI xs)
```

Comment:
Optional thought: This logic seems like the write logic for serializing any `NP`. Do we already have an instance for that? `... => ToCBOR (NP f xs)` and `... => FromCBOR (NP f xs)` eg?

Reply:
I think this implementation is actually wrong regarding versioning. It should omit elements for disabled eras as is done in `SerialiseNodeToClient (HardForkBlock xs) (PerEraLedgerConfig xs)`. I'll try to generalize an encoder/decoder for `NP f xs`.

Reply:
Oh my! That turned out to be a bit of a headache, but I think I managed. In the end I had to:

* Move `SerialiseNodeToClient (HardForkBlock xs) (HardForkLedgerConfig xs)` from `Ouroboros.Consensus.HardFork.Combinator.Basics` to `Ouroboros.Consensus.HardFork.Combinator.Serialisation.SerialiseNodeToClient` (I still dislike orphan instances, but with all the module cycle that I had to deal with, I finally gave in :shrug: ).
* Added a generalized encoding/decoding for `NP f xs`, namely `encodeNodeToClientNP` and `encodeNodeToClientNP` and use it to implemented `SerialiseNodeToClient` instances for `Shape xs` and `PerEraLedgerConfig xs`

I'll push these change after I respond to the remaining comments.